File: agent/typescript/src/protocol.ts
export type AgentMessageRole = "system" | "user" | "assistant";

export type AgentRunStatus =
  | "idle"
  | "running"
  | "waiting_for_approval"
  | "completed"
  | "failed"
  | "cancelled";

export type AgentRunMode = "chat" | "plan" | "edit";

export type AgentApiStyle = "openai_compatible" | "anthropic_compatible";

export type AgentSearchMode = "auto" | "disabled" | "tavily";

export type AgentReadPermission = "workspace_only" | "all";

export type AgentWritePermission = "denied" | "workspace_only" | "all";

export type AgentCommandPermission = "require_approval" | "auto_approve";

export type AgentPatchPermission = "require_approval" | "auto_approve";

export interface AgentPermissions {
  read: AgentReadPermission;
  write: AgentWritePermission;
  command: AgentCommandPermission;
  patch: AgentPatchPermission;
}

export type AgentPromptWorkMode = "coding" | "general";

export type AgentPromptTone = "friendly" | "pragmatic";

export type AgentPromptDetailLevel = "low" | "medium" | "high";

export type AgentToolName =
  | "attachments_list"
  | "attachments_list_project"
  | "read_file"
  | "read_image"
  | "read_pdf"
  | "read_word"
  | "read_presentation"
  | "read_spreadsheet"
  | "workspace_map"
  | "search_files"
  | "search_code"
  | "web_search"
  | "web_fetch"
  | "git_diff"
  | "apply_patch"
  | "run_command"
  | (string & {});

export type AgentToolSafety = "read_only" | "requires_approval" | "destructive";

export type AgentApprovalStatus = "not_required" | "required" | "approved" | "rejected";

export type AgentApprovalDecisionStatus = "approved" | "rejected";

export type AgentPatchOperation = "create" | "update" | "delete";

export type AgentPatchResultStatus = "applied" | "failed" | "rejected";

export type AgentCommandOutputStream = "stdout" | "stderr";

export type AgentCommandRiskLevel =
  | "read_only"
  | "writes_workspace"
  | "network"
  | "destructive"
  | "unknown";

export interface AgentChatMessage {
  role: AgentMessageRole;
  content: string;
}

export type AgentInputAttachmentKind = "file" | "image";

export type AgentInputAttachmentEncoding = "utf8" | "base64";

export interface AgentInputAttachment {
  id: string;
  kind: AgentInputAttachmentKind;
  name: string;
  mimeType?: string;
  sizeBytes: number;
  encoding: AgentInputAttachmentEncoding;
  data: string;
  truncated?: boolean;
}

export interface AgentConversationMessageAttachment {
  id: string;
  kind: AgentInputAttachmentKind;
  name: string;
  mimeType?: string | null;
  sizeBytes: number;
  previewData?: string | null;
  previewMimeType?: string | null;
  createdAt?: number;
}

export interface AgentWorkspaceContext {
  projectId?: string;
  displayName?: string;
  rootPath?: string;
}

export interface AgentAttachmentReference {
  id: string;
  conversationId: string;
  messageId: string;
  projectId?: string | null;
  kind: AgentInputAttachmentKind;
  name: string;
  mimeType?: string;
  sizeBytes: number;
  readPath: string;
  storageRelPath: string;
  createdAt: number;
}

export interface AgentAttachmentLibraryContext {
  rootPath?: string;
  conversationId?: string;
  projectId?: string | null;
  conversationAttachments: AgentAttachmentReference[];
  projectAttachments: AgentAttachmentReference[];
}

export interface AgentPromptPreferences {
  workMode?: AgentPromptWorkMode;
  tone?: AgentPromptTone;
  detailLevel?: AgentPromptDetailLevel;
  customInstructions?: string;
  updatedAt?: number;
}

export interface AgentRunContext {
  conversationId?: string;
  projectId?: string | null;
  workspace?: AgentWorkspaceContext;
  attachmentLibrary?: AgentAttachmentLibraryContext;
  permissions: AgentPermissions;
}

export interface AgentSearchConfig {
  mode: AgentSearchMode;
  tavilyApiKey?: string;
}

export interface AgentApprovalDecision {
  actionId: string;
  status: AgentApprovalDecisionStatus;
  message?: string;
}

export interface AgentUsage {
  inputTokens?: number;
  outputTokens?: number;
  totalTokens?: number;
  cachedInputTokens?: number;
  cacheCreationInputTokens?: number;
  billableRequestCount?: number;
}

export type AgentUsageSummaryRange = "last7Days" | "last30Days" | "all" | "custom";

export interface AgentUsageSummaryInput {
  range: AgentUsageSummaryRange;
  from?: number;
  to?: number;
}

export interface AgentUsageModelSummary {
  modelId: string;
  modelName: string;
  providerPath?: string;
  requestCount: number;
  messageCount: number;
  inputTokens?: number;
  outputTokens?: number;
  totalTokens?: number;
  cachedInputTokens?: number;
  cacheCreationInputTokens?: number;
  estimatedCost?: number;
}

export interface AgentUsageSummaryOutput {
  requestCount: number;
  messageCount: number;
  inputTokens?: number;
  outputTokens?: number;
  totalTokens?: number;
  cachedInputTokens?: number;
  cacheCreationInputTokens?: number;
  estimatedCost?: number;
  models: AgentUsageModelSummary[];
}

export interface AgentUsageClearInput {
  from?: number;
  to?: number;
}

export interface AgentUsageClearOutput {
  deletedRecords: number;
}

export interface AgentChatOutput {
  status: AgentRunStatus;
  content: string;
  runId: string;
  events: AgentEvent[];
  toolDefinitions: AgentToolDefinition[];
  usage?: AgentUsage;
  finishReason?: string;
  proposedActions: AgentProposedAction[];
}

export interface AgentConversationTurnInput {
  conversationId?: string;
  projectId?: string | null;
  modelId: string;
  content: string;
  attachments?: AgentInputAttachment[];
  title?: string;
  userMessageId?: string;
  assistantMessageId?: string;
  maxTokens?: number;
  temperature?: number;
  mode?: AgentRunMode;
  promptPreferences?: AgentPromptPreferences;
  permissions?: AgentPermissions;
}

export interface AgentConversationMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  createdAt: number;
  status?: "pending" | "sent" | "error" | null;
  attachments?: AgentConversationMessageAttachment[];
}

export interface AgentConversationTurnOutput {
  runId: string;
  eventName: string;
  conversationId: string;
  userMessageId: string;
  assistantMessageId: string;
  userMessage: AgentConversationMessage;
  assistantMessage: AgentConversationMessage;
}

export interface AgentStateSnapshot {
  status: AgentRunStatus;
  activeRunId: string | null;
  lastError: string | null;
  updatedAt: number;
}

export interface AgentToolCall {
  id: string;
  tool: AgentToolName;
  args: unknown;
  approvalStatus: AgentApprovalStatus;
  reason?: string;
}

export interface AgentToolDefinition {
  name: AgentToolName;
  description: string;
  inputSchema: unknown;
  safety: AgentToolSafety;
  requiresWorkspace: boolean;
  requiresApproval: boolean;
}

export interface AgentToolResult {
  callId: string;
  tool: AgentToolName;
  ok: boolean;
  result?: unknown;
  error?: string;
}

export interface AgentToolContinuation {
  call: AgentToolCall;
  result: AgentToolResult;
}

export interface AgentDiffProposal {
  id: string;
  operation: AgentPatchOperation;
  filePath: string;
  patch: string;
  baseRevision?: string;
  summary?: string;
  approvalStatus: AgentApprovalStatus;
}

export interface AgentPatchResult {
  status: AgentPatchResultStatus;
  operation: AgentPatchOperation;
  filePath: string;
  appliedFilePaths: string[];
  gitDiff?: AgentGitDiffSnapshot;
  gitDiffError?: string;
  error?: string;
  message?: string;
}

export interface AgentGitDiffSnapshot {
  patch: string;
  truncated: boolean;
}

export interface AgentCommandRequest {
  id: string;
  command: string;
  cwd?: string;
  timeoutMs?: number;
  approvalStatus: AgentApprovalStatus;
  riskLevel?: AgentCommandRiskLevel;
  reason?: string;
}

export type AgentProposedAction =
  | { type: "tool_call"; call: AgentToolCall }
  | { type: "diff"; diff: AgentDiffProposal }
  | { type: "command"; command: AgentCommandRequest };

export type AgentEvent =
  | { type: "started"; runId: string; toolDefinitions: AgentToolDefinition[] }
  | { type: "state"; runId: string; state: AgentStateSnapshot }
  | { type: "message_delta"; runId: string; delta: string }
  | { type: "message"; runId: string; content: string }
  | { type: "tool_call"; runId: string; call: AgentToolCall }
  | { type: "tool_result"; runId: string; result: AgentToolResult }
  | { type: "approval_required"; runId: string; action: AgentProposedAction }
  | { type: "diff"; runId: string; diff: AgentDiffProposal }
  | {
      type: "command_output";
      runId: string;
      command: string;
      stream: AgentCommandOutputStream;
      output: string;
    }
  | { type: "error"; runId?: string; message: string; recoverable: boolean }
  | {
      type: "done";
      runId: string;
      success: boolean;
      status?: AgentRunStatus;
      content?: string;
      usage?: AgentUsage;
      finishReason?: string;
      proposedActions?: AgentProposedAction[];
    };

File: agent/rust/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::error::Error;
use std::fmt::{Display, Formatter};

#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentChatInput {
    pub api_url: String,
    pub api_token: String,
    pub model: String,
    pub api_style: Option<AgentApiStyle>,
    pub max_tokens: Option<u32>,
    pub temperature: Option<f32>,
    pub mode: Option<AgentRunMode>,
    pub stream: Option<bool>,
    pub context: Option<AgentRunContext>,
    pub search_config: Option<AgentSearchConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_preferences: Option<AgentPromptPreferences>,
    pub approval_decision: Option<AgentApprovalDecision>,
    #[serde(default)]
    pub attachments: Vec<AgentInputAttachment>,
    pub messages: Vec<AgentChatMessage>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct AgentChatMessage {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentInputAttachmentKind {
    File,
    Image,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentInputAttachmentEncoding {
    Utf8,
    Base64,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentInputAttachment {
    pub id: String,
    pub kind: AgentInputAttachmentKind,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    pub size_bytes: u64,
    pub encoding: AgentInputAttachmentEncoding,
    pub data: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub truncated: Option<bool>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentChatOutput {
    pub content: String,
    pub status: AgentRunStatus,
    pub run_id: String,
    pub events: Vec<AgentEvent>,
    pub tool_definitions: Vec<AgentToolDefinition>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<AgentUsage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub finish_reason: Option<String>,
    pub proposed_actions: Vec<AgentProposedAction>,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentRunStatus {
    Idle,
    Running,
    WaitingForApproval,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentRunMode {
    Chat,
    Plan,
    Edit,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentApiStyle {
    OpenAiCompatible,
    AnthropicCompatible,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentSearchMode {
    Auto,
    Disabled,
    Tavily,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentPromptWorkMode {
    Coding,
    General,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentPromptTone {
    Friendly,
    Pragmatic,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentPromptDetailLevel {
    Low,
    Medium,
    High,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentPromptPreferences {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub work_mode: Option<AgentPromptWorkMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tone: Option<AgentPromptTone>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail_level: Option<AgentPromptDetailLevel>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_instructions: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<i64>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentSearchConfig {
    pub mode: AgentSearchMode,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tavily_api_key: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentApprovalDecision {
    pub action_id: String,
    pub status: AgentApprovalDecisionStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentRunContext {
    pub conversation_id: Option<String>,
    pub project_id: Option<String>,
    pub workspace: Option<AgentWorkspaceContext>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attachment_library: Option<AgentAttachmentLibraryContext>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentWorkspaceContext {
    pub project_id: Option<String>,
    pub display_name: Option<String>,
    pub root_path: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentAttachmentLibraryContext {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub root_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    #[serde(default)]
    pub conversation_attachments: Vec<AgentAttachmentReference>,
    #[serde(default)]
    pub project_attachments: Vec<AgentAttachmentReference>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentAttachmentReference {
    pub id: String,
    pub conversation_id: String,
    pub message_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    pub kind: AgentInputAttachmentKind,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    pub size_bytes: u64,
    pub read_path: String,
    pub storage_rel_path: String,
    pub created_at: i64,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentUsage {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_tokens: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_tokens: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_tokens: Option<u64>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentStateSnapshot {
    pub status: AgentRunStatus,
    pub active_run_id: Option<String>,
    pub last_error: Option<String>,
    pub updated_at: u64,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentApprovalStatus {
    NotRequired,
    Required,
    Approved,
    Rejected,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentApprovalDecisionStatus {
    Approved,
    Rejected,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentToolSafety {
    ReadOnly,
    RequiresApproval,
    Destructive,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentCommandOutputStream {
    Stdout,
    Stderr,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentCommandRiskLevel {
    ReadOnly,
    WritesWorkspace,
    Network,
    Destructive,
    Unknown,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentToolCall {
    pub id: String,
    pub tool: String,
    pub args: Value,
    pub approval_status: AgentApprovalStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentToolDefinition {
    pub name: String,
    pub description: String,
    pub input_schema: Value,
    pub safety: AgentToolSafety,
    pub requires_workspace: bool,
    pub requires_approval: bool,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentToolResult {
    pub call_id: String,
    pub tool: String,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentDiffProposal {
    pub id: String,
    pub file_path: String,
    pub patch: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    pub approval_status: AgentApprovalStatus,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AgentCommandRequest {
    pub id: String,
    pub command: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout_ms: Option<u64>,
    pub approval_status: AgentApprovalStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub risk_level: Option<AgentCommandRiskLevel>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum AgentProposedAction {
    ToolCall { call: AgentToolCall },
    Diff { diff: AgentDiffProposal },
    Command { command: AgentCommandRequest },
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum AgentEvent {
    Started {
        run_id: String,
        tool_definitions: Vec<AgentToolDefinition>,
    },
    State {
        run_id: String,
        state: AgentStateSnapshot,
    },
    MessageDelta {
        run_id: String,
        delta: String,
    },
    Message {
        run_id: String,
        content: String,
    },
    ToolCall {
        run_id: String,
        call: AgentToolCall,
    },
    ToolResult {
        run_id: String,
        result: AgentToolResult,
    },
    ApprovalRequired {
        run_id: String,
        action: AgentProposedAction,
    },
    Diff {
        run_id: String,
        diff: AgentDiffProposal,
    },
    CommandOutput {
        run_id: String,
        command: String,
        stream: AgentCommandOutputStream,
        output: String,
    },
    Error {
        run_id: Option<String>,
        message: String,
        recoverable: bool,
    },
    Done {
        run_id: String,
        success: bool,
        #[serde(skip_serializing_if = "Option::is_none")]
        status: Option<AgentRunStatus>,
        #[serde(skip_serializing_if = "Option::is_none")]
        content: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        usage: Option<AgentUsage>,
        #[serde(skip_serializing_if = "Option::is_none")]
        finish_reason: Option<String>,
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        proposed_actions: Vec<AgentProposedAction>,
    },
}

#[derive(Debug, Clone)]
pub struct AgentError {
    message: String,
}

pub type AgentResult<T> = Result<T, AgentError>;

impl AgentError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl Display for AgentError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl Error for AgentError {}

impl From<String> for AgentError {
    fn from(message: String) -> Self {
        Self::new(message)
    }
}

impl From<&str> for AgentError {
    fn from(message: &str) -> Self {
        Self::new(message)
    }
}
